nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py

```diff
@@ -34,11 +75,57 @@ def __init__(self, dim: int, seq_len_interpolation_factor: int = None):
-        inv_freq = 1.0 / (10000 ** (torch.arange(0, dim, 2).float() / dim))
-        self.register_buffer('inv_freq', inv_freq)
+        if self.use_yarn:
+            self.seq_len_interpolation_factor = seq_len_interpolation_factor  # scale
+            self.pretrain_max_positional_embeddings = pretrain_max_positional_embeddings
+            self.extrapolation_factor = extrapolation_factor
+            self.attn_factor = attn_factor
+            self.beta_fast = beta_fast
+            self.beta_slow = beta_slow
+            self.dim = dim
+            pos_freqs = self.base ** (torch.arange(0, self.dim, 2).float() / self.dim)
+            inv_freq_extrapolation = 1.0 / pos_freqs
+            inv_freq_interpolation = 1.0 / (self.seq_len_interpolation_factor * pos_freqs)
+
+            low, high = find_correction_range(
+                self.beta_fast, self.beta_slow, self.dim, self.base, self.pretrain_max_positional_embeddings
+            )
+            inv_freq_mask = (
+                1 - linear_ramp_mask(low, high, self.dim // 2).float()
+            ) * self.extrapolation_factor  # Get n-d rotational scaling corrected for extrapolation
+            inv_freq = inv_freq_interpolation * (1 - inv_freq_mask) + inv_freq_extrapolation * inv_freq_mask
+
+            self.register_buffer("inv_freq", inv_freq)
+            self.mscale = float(
+                get_mscale(self.seq_len_interpolation_factor) * self.attn_factor
+            )  # Get n-d magnitude scaling corrected for interpolation
+
+            self.max_seq_len_cached = max_position_embeddings
+            t = torch.arange(self.max_seq_len_cached, device=self.inv_freq.device, dtype=self.inv_freq.dtype)
+            freqs = torch.einsum("i , j -> i j", t, self.inv_freq)
+            # Different from paper, but it uses a different permutation in order to obtain the same calculation
+            emb = torch.cat((freqs, freqs), dim=-1)
+            self.register_buffer('emb', emb)
+
+        else:
+            inv_freq = 1.0 / (self.base ** (torch.arange(0, dim, 2).float() / dim))
+            self.register_buffer('inv_freq', inv_freq)
 
     def forward(self, max_seq_len, offset=0):
+        if self.use_yarn:
+            if max_seq_len > self.max_seq_len_cached:
+                self.max_seq_len_cached = max_seq_len
+
+                t = torch.arange(self.max_seq_len_cached, dtype=self.inv_freq.dtype).to(self.inv_freq.device)
+                freqs = torch.einsum("i , j -> i j", t, self.inv_freq)
+                # Different from paper, but it uses a different permutation in order to obtain the same calculation
+                emb = torch.cat((freqs, freqs), dim=-1).to(self.inv_freq.device)
+                self.register_buffer('emb', emb)
+            return rearrange(self.emb, 'n d -> n 1 1 d')
+
         seq = torch.arange(max_seq_len, device=self.inv_freq.device) + offset
         if self.seq_len_interpolation_factor is not None:
```

Comment:
We can simply add `if self.seq_len_interpolation_factor is not None and not self.use_yarn:`